revocation/ocsp/ocsp.go

```diff
@@ -0,0 +1,180 @@
+// Package OCSP provides methods for checking the OCSP revocation status of a certificate chain
+// as well as errors related to these checks
+package ocsp
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/x509"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"golang.org/x/crypto/ocsp"
+)
+
+type Options struct {
+	CertChain       []*x509.Certificate
+	SigningTime     time.Time
+	HTTPClient      *http.Client
+	MergeErrorsFunc func(errors []error) error
+}
+
+const pkixNoCheckOID string = "1.3.6.1.5.5.7.48.1.5"
+const ocspMaxResponseSize int64 = 20480 //bytes
+
+func OCSPStatus(opts Options) error {
+	for i, cert := range opts.CertChain {
+		if i != (len(opts.CertChain) - 1) {
+			if err := validateIssuer(cert, opts.CertChain[i+1]); err != nil {
+				return CheckOCSPError{Err: errors.New("invalid chain: expected chain to be correct and complete with each cert issued by the next in the chain")}
+			}
+		} else {
+			if err := validateIssuer(cert, cert); err != nil {
+				return CheckOCSPError{Err: errors.New("invalid chain: expected chain to end with root cert")}
+			}
+		}
+	}
+
+	wg := sync.WaitGroup{}
+	certResults := make([]error, len(opts.CertChain))
+	for i, cert := range opts.CertChain {
+		if i != (len(opts.CertChain) - 1) {
+			wg.Add(1)
+			// Assume cert chain is accurate and next cert in chain is the issuer
+			go certOCSPStatus(cert, opts.CertChain[i+1], &certResults[i], &wg, opts)
```

Comment:
Just an open discussion, what is the reason that we choose multiple `goroutines` here? Will a single-threaded design also work?

Reply:
Single threaded would also work, but would be much slower for larger chains. Parallelization allows us to perform checks on multiple certs simultaneously. The main reason is that there may be multiple HTTP requests for each cert in the list. Since the HTTP requests may be slow, it is best not to wait for them to start other unrelated calls (hence the multiple goroutines).

Take this example, if you have a cert chain that has 8 certificates that each have 2 OCSP servers, and for some reason they all time out. With a 5 second timeout, that would be almost 1.5 minutes to wait for the result if it is single threaded. However, if we multithread on an 8 core system, it would drop to only 10 seconds for that same result.